### ISSUE TITLE
[SR-15368] ProcessInfo.processInfo.hostName not thread-safe, can lead to crashes under Linux

### ISSUE THREAD

**Issue by weissi** [apple/swift-corelibs-foundation#3194]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15368      |
   |Radar             | rdar://problem/84517827         |
   |Original Reporter | @weissi      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Foundation    |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: ce7ed0ffe39e9c016347978e5e6bd8db

  </details>

**Issue Description:**

`ProcessInfo` is documented to be thread-safe. Unfortunately, on Linux it isn't and calling `ProcessInfo.processInfo.hostName` may actually lead to crashes.

The problem is that `ProcessInfo` calls `Host.current().name` (<https://github.com/apple/swift-corelibs-foundation/blob/main/Sources/Foundation/ProcessInfo.swift#L61)>

`Host.name` then calls `Host.names` (<https://github.com/apple/swift-corelibs-foundation/blob/main/Sources/Foundation/Host.swift#L335)> which is mutating the instance without any synchronisation on the singleton `Host._current`.

For example, if you run this program in a loop, you'll see a segfault (or some other crash) after a while)

    import Foundation
    import Dispatch

    let threads = 60
    let sem1 = DispatchSemaphore(value: 0)
    let sem2 = DispatchSemaphore(value: 0)
    let g = DispatchGroup()

    (0..<threads).forEach { index in
        DispatchQueue(label: "\(index)").async(group: g) {
            sem1.signal()
            // All threads should be parked here
            sem2.wait()
            precondition(ProcessInfo.processInfo.hostName.utf8.count > 0)
        }
    }

    (0..<threads).forEach { index in
        sem1.wait()
    }
    (0..<threads).forEach { index in
        sem2.signal()
    }
    g.wait()

    $ docker run -it --rm -v "$PWD:$PWD" -w "$PWD" swift:5.5-focal bash -c 'swiftc repro.swift && while ./repro; do echo -n .; done'
    .............bash: line 1:   833 Segmentation fault      ./repro

**Comment by weissi** [comment]
and as well on today's `main`

```
$ jw-docker-swift-main bash -c 'swiftc repro.swift && while ./repro; do echo -n .; done'
docker.io/swiftlang/swift:nightly-main
...........bash: line 1:   710 Segmentation fault      ./repro
```

**Comment by weissi** [comment]
This is fixed in `swift-foundation` but not in this repo.

**Comment by parkera** [comment]
@weissi do you mean that we didn't actually end up removing the duplicate ProcessInfo in swift-corelibs-foundation?

**Comment by weissi** [comment]
> @weissi do you mean that we didn't actually end up removing the duplicate ProcessInfo in swift-corelibs-foundation?

Sorry, no, I think this issue is fixed in Swift 6.0. So I think we can close this

**Comment by weissi** [comment]
> > @weissi do you mean that we didn't actually end up removing the duplicate ProcessInfo in swift-corelibs-foundation?
> 
> Sorry, no, I think this issue is fixed in Swift 6.0. So I think we can close this.
> 
> But I believe the old busted code is still in this repo, let me look.

actually, it seems fine. @parkera I think we should close this
